rust_bindings/src/lib.rs

```diff
@@ -0,0 +1,131 @@
+        if ptr.is_null() {
+            return None;
+        }
+        let c_str = CStr::from_ptr(ptr).to_owned();
+        bindings::free(ptr as *mut std::ffi::c_void);
+        Some(c_str)
+    }
+}
+
+pub fn verify_n_wesolowski(
+    discriminant: &CStr,
+    x_s: &[u8],
+    proof: &[u8],
+    num_iterations: u64,
+    disc_size_bits: u64,
+    recursion: u64,
+) -> bool {
+    // SAFETY: The lengths are guaranteed to match the actual lengths of the char pointers.
+    unsafe {
+        let value = bindings::verify_n_wesolowski_wrapper(
+            discriminant.as_ptr() as *const std::ffi::c_char,
+            x_s.as_ptr() as *const std::ffi::c_char,
+            x_s.len(),
+            proof.as_ptr() as *const std::ffi::c_char,
+            proof.len(),
+            num_iterations,
+            disc_size_bits,
+            recursion,
+        );
+        value == 1
+    }
+}
+
+pub fn prove(
+    challenge: &[u8],
+    x_s: &[u8],
+    disc_size_bits: i32,
+    num_iterations: u64,
+) -> Option<Vec<u8>> {
+    // SAFETY: The lengths are guaranteed to match the actual lengths of the char pointers.
+    unsafe {
+        let array = bindings::prove_wrapper(
+            challenge.as_ptr(),
+            challenge.len(),
+            x_s.as_ptr(),
+            x_s.len(),
+            disc_size_bits,
+            num_iterations,
+        );
+        if array.data.is_null() {
+            return None;
+        }
+        let result = std::slice::from_raw_parts(array.data, array.length).to_vec();
+        bindings::delete_byte_array(array);
+        Some(result)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use hex_literal::hex;
+    use rand::{Rng, SeedableRng};
+    use rand_chacha::ChaCha8Rng;
+
+    use super::*;
+
+    #[test]
+    fn test_create_discriminant() {
+        let mut discriminants = Vec::new();
+        let mut seeds = [[0; 10]; 5];
+
+        for (i, seed) in seeds.iter_mut().enumerate() {
+            let mut rng = ChaCha8Rng::seed_from_u64(i as u64);
+
+            rng.fill(seed);
+
+            let discriminant = create_discriminant(seed, 512).unwrap();
+
+            discriminants.push(discriminant);
+        }
+
+        let expected = [
+            "-0x9a8eaf9c52d9a5f1db648cdf7bcd04b35cb1ac4f421c978fa61fe1344b97d4199dbff700d24e7cfc0b785e4b8b8023dc49f0e90227f74f54234032ac3381879f",
+            "-0xb193cdb02f1c2615a257b98933ee0d24157ac5f8c46774d5d635022e6e6bd3f7372898066c2a40fa211d1df8c45cb95c02e36ef878bc67325473d9c0bb34b047",
+            "-0xbb5bd19ae50efe98b5ac56c69453a95e92dc16bb4b2824e73b39b9db0a077fa33fc2e775958af14f675a071bf53f1c22f90ccbd456e2291276951830dba9dcaf",
+            "-0xa1e93b8f2e9b0fd3b1325fbe40601f55e2afbdc6161409c0aff8737b7213d7d71cab21ffc83a0b6d5bdeee2fdcbbb34fbc8fc0b439915075afa9ffac8bb1b337",
+            "-0xf2a10f70148fb30e4a16c4eda44cc0f9917cb9c2d460926d59a408318472e2cfd597193aa58e1fdccc6ae6a4d85bc9b27f77567ebe94fcedbf530a60ff709fd7",
+        ];
```

Comment:
where do these test vectors come from? I think it would be good to have a comment here with a reference

Reply:
I added a comment and the actual seeds in the code instead of the RNG, so you can easily compare with Python if needed.